### DISCUSSION TITLE
Last.fm Scrobbling

### DISCUSSION THREAD

**Discussion by mdrxy** [AzuraCast/AzuraCast#7566] [Ideas]
Hello,

I am hoping to extend AzuraCast to allow stations to scrobble songs played to their Last.fm profiles. 

Before I get into the implementation, there are a few questions I have:

1. Would this make more sense under station settings or as a webhook option?
2. Given that not all songs will be played in full (e.g. manually triggering a song skip or a playlist with higher priority beginning), it probably doesn't make sense to complete a scrobble only following a complete playback. I think some kind of slider, such as "Scrobble after x% of playback" would help to address this and give station owners the opportunity to tune this to their liking. Could someone walk me through how I might be able to trigger this event once the timestamp correlating to 50% completion, for instance, is reached? Or is this even possible under the current architecture?
3. Under what logging category would this feature best fit within?
4. Are there any other edge cases I am missing / should consider?

Much thanks,
Mason

**Comment by gAlleb** [comment]
Hi! 

Liquidsoap already has a native `last.fm` extension builtin. To use it all you have to do is just add a tiny bit of code to your config.

Put this into 3rd box of LS config with `username` and `password` and you are good to go:

```ruby
def lastfm.submit(~user,~password,~source="broadcast",~length=false) = 

fun (m) ->

if (m["jingle_mode"] != "true") then
  audioscrobbler.submit(user=user,password=password,
                source=source,length=length,
                host="post.audioscrobbler.com",port=80,
                m)
end

end

radio.on_metadata(lastfm.submit(user="USER",password="PASS"))
```

That's it.

Now it avoids jingles and sends data to last.fm.

p.s. with some basic `if`, `then`, `thread.run/delay` manipulations you can fine tune it to do any advanced stuff you need. By default it works very fine too. 

https://github.com/savonet/liquidsoap/blob/ef2a2dbfc2cf5c3fc52a5a27f468b8548679af49/src/libs/extra/lastfm.liq

**Reply by mdrxy** [reply]
This is good to know - thank you. This seems like an excellent option for stations who use Liquidsoap as their source. For others, however, it looks like a different approach will be needed. For instance, our stream relies entirely on Icecast metadata updates via:

```
/admin/metadata?mount=/stream&mode=updinfo&artist={artist name}&title={song name}
```

To achieve scrobbling functionality we'd likely need to plug in to the logic that updates the song history. I'll see what I can find.

**Comment by BusterNeece** [comment]
If you were to add this to AzuraCast, I'd suspect a webhook would be the best place to put it. It meets all the expectations of a webhook in that it would dispatch a remote URL upon track changes.

It would be a lot trickier to make it wait for a track to play through completely, so I'd likely just have it scrobble right as the track is indicated as having started playing.

**Comment by LordFader** [comment]
It work in the last Stable Release...

I try it in the Rolling Release #[cd70097] and isn't working

**Reply by gAlleb** [reply]
Yes. There is a bug introduced in latest Liquidsoap. It will be addressed soon.

**Comment by itzboterberg** [comment]
hello i wont this also but where do i add this?

**Reply by gAlleb** [reply]
1. https://github.com/AzuraCast/AzuraCast/discussions/7566#discussioncomment-11327747

2. https://github.com/AzuraCast/AzuraCast/discussions/7566#discussioncomment-11476822

**Comment by itzboterberg** [comment]
okay i found it where to place but now last.fm send the music two time i found someyhing but evrey time i try it this is the error
`At line 78, char 44-44: Error 5: this value has type source(_) .{ buffered : _, duration : _, elapsed : _, fallible : _, id : _, is_active : _, is_ready : _, is_up : _, last_metadata : _, log : _, on_metadata : _, on_shutdown : _, on_track : _, on_wake_up : _, remaining : _, seek : _, self_sync : _, skip : _, time : _ } but it should be a subtype of { buffered? : _, duration? : _, elapsed? : _, fallible? : _, id? : _, is_active? : _, is_ready? : _, is_up? : _, last_metadata? : _, log? : _, on_metadata? : _, on_shutdown? : _, on_track? : _, on_wake_up? : _, remaining? : _, seek? : _, self_sync? : _, skip? : _, time? : _ } (inferred at extra/lastfm.liq, line 85, char 2-32)`  this is the code 

```lua
def lastfm.submit.full(~user, ~password, ~source, ~length, ~delay, ~force, ~s) =
    fun (m) -> 
        if (m["jingle_mode"] != "true") then
            audioscrobbler.submit.full(user=user, password=password,
                source=source, length=length,
                host="post.audioscrobbler.com", port=80,
                delay=delay, force=force, s)
        end
    end

    radio.on_metadata(lastfm.submit.full(
      user="USERNAME",
      password="PASSWORD",
      source="broadcast",
      length=false,
      delay=10.,
      force=false,
      s=some_value
  ))
```

**Reply by gAlleb** [reply]
You are using wrong script.

Please follow this guide: https://www.azuracast.com/docs/developers/liquidsoap/#scrobbling-to-lastfm

And beware there is a bug introduced in latest Liquidsoap (and latest AzuraCCast Rolling Release) regarding scrobbling. It will be addressed soon.

**Reply by itzboterberg** [reply]
yeah but with that script it upload two times the music
https://servwave.org/rstore/rxpo9vqt78n9aqqy97oddzi9vpj2jnakt0kxofti.png

**Reply by gAlleb** [reply]
Have you put in the 3rd box?

**Reply by itzboterberg** [reply]
yes

**Reply by itzboterberg** [reply]
i have put it under # Handle "Jingle Mode" tracks by replaying the previous metadata.
radio = azuracast.handle_jingle_mode(radio)

**Reply by gAlleb** [reply]
Then I don't know what is happening.  What version of Azuracast you are on?

**Reply by itzboterberg** [reply]
v0.20.4 Stable

**Reply by gAlleb** [reply]
Why have you put it there? Please follow he instructions and put it into the 3rd box:
THIRD BOX!

It must be BEFORE `radio = azuracast.handle_jingle_mode(radio)` or it will double metadata.

**Reply by itzboterberg** [reply]
ow okay thank you

**Reply by LordFader** [reply]
> yeah but with that script it upload two times the music https://servwave.org/rstore/rxpo9vqt78n9aqqy97oddzi9vpj2jnakt0kxofti.png

Maybe you are scrobbling in duplicate... from your browser or audio app

and now with your script in your radio...

**Comment by gAlleb** [comment]
@mdrxy Hi!

> Liquidsoap has a native [Last.fm](https://www.last.fm/) extension builtin. To use it, you just need to add some custom code to your Liquidsoap config in the third text box (underneath add_skip_command(radio)).

You've added `(underneath add_skip_command(radio)).` instead of `(before add_skip_command(radio)).`. I think it might confuse some people)

**Comment by BusterNeece** [comment]
Definitely going to work on integrating this as a web hook. I don't want such a heavily-used feature to be buried in advanced code instructions, as it clearly causes confusion.

**Reply by gAlleb** [reply]
There is also nice "full" function which is applied directly to source:
`radio = lastfm.submit.full(user="USERNAME", password="PASSWORD", radio)`
It shows `now playing` during playback and scrobbles when track reaches the end. But it will show and scrobble jingles also.

**Comment by gAlleb** [comment]
### Behold!

New `last.fm scrobbling` functionality `reimplemented from scratch in Liquidsoap language`. 
It uses `new` last.fm protocol and will **`soon`** be available in Liquidsoap Rolling Release. 
https://github.com/savonet/liquidsoap/blob/main/src/libs/extra/audioscrobbler.liq
<sup>Old `lastfm...` functions will be removed.</sup>

<hr/>

### If you'd like to fire this up when it's ready:

1. Get an API account to obtain "API Key" and "API Secret" (if you don't have them already): https://www.last.fm/api/account/create (You can check the created API applications here: https://www.last.fm/api/accounts)
2. You might want to disconnect your application to get new session key here: https://www.last.fm/settings/applications. It's not always necessary, just in case.
3. If you have any old last.fm related code in your LS config - remove it.
4. Use your credentials (`API_KEY, API_SECRET, USERNAME, PASSWORD`) and paste the following into 3rd box of LS config:

```ruby
settings.audioscrobbler.api_key := "API_KEY"
settings.audioscrobbler.api_secret := "API_SECRET"

radio = audioscrobbler.submit(
  username="USERNAME",
  password="PASSWORD",
  force=true, 
  metadata_preprocessor=(fun(m) -> if m["jingle_mode"] == "true" then [] else m end),
 radio
)
```

That's it. _The One small code to rule them all._

- While track is being played it's being marked as `Scrobbling now` on last.fm page. 
- When track finishes - it becomes scrobbled. 
- Jingles are not scrobbled. 
- Tracks with no `Artist` or no `Title` are not scrobbled.

<hr/>

P.S.

1. `force=true,` -  is optional line.  If remaining time is null, the song will be assumed to be skipped or cut, and not submitted. Setting `force=true` prevents this behavior.
2. `metadata_preprocessor=` - is optional line. Setting `metadata_preprocessor=(fun(m) -> if m["jingle_mode"] == "true" then [] else m end),` will prevent jingles from scrobbling. If function retunrns `[]` - scrobbling is skipped.
3. <sup>New audioscrobbler operators are added, most importantly `audioscrobbler.api.track.updateNowPlaying` and `audioscrobbler.api.track.scrobble`. The operators are combined into `audioscrobbler.submit` which takes a source and applies `updateNowPlaying` when a track starts and `scrobble` when it ends.</sup>

**Reply by userwithroot** [reply]
On the latest version of AzuraCast, this and last method does not work :(

**Reply by gAlleb** [reply]
> It uses new last.fm protocol and `will soon be available` in Liquidsoap Rolling Release.

**Comment by paulcjkp** [comment]
I've been using this for a few days now, works great, except one tiny issue.  The band 311 scrobbles as 311.00

I have a few songs that are just numbers, the .00 gets added to those as well.  I'm not sure how songs are being sent to Last.fm, but it seems like the text is being converted to a number with 2 decimal places.

**Reply by gAlleb** [reply]
Hi @paulcjkp, I can't reproduce this behavior. Could you share a file that causes this?

**Reply by paulcjkp** [reply]
https://we.tl/t-K3zLkhlUs8

2 files causing issues.

**Reply by gAlleb** [reply]
Thank you, I've tested before with bare Liquidsoap and now have tested with the latest RR - https://github.com/AzuraCast/AzuraCast/commit/f24c5168b1c87934b683145171b2bbfa0d541835

Yes, I can reproduce this bug only on AzuraCast instance. This is probably caused by how annotations come into Azuracast. Probably you should see something like this too:

```
2025/01/09 14:29:43 [azuracast.api:3] API nextsong - Response (200): 
{"uri":"annotate:title=\"22.00\",artist=\"Bartek\",duration=\"151.01\",
song_id=\"61a80a834d7f587ace9dde8d823d850c\",media_id=\"33940\",sq_id=\"222465\",playlist_id=\"68\":
media:music_local/test/02_-_bartek_-_22.mp3"}
```

Do you have similar annotations?

Somehow if an artist/title string is a number it converts it to a float. And because this meta is sent before other AC functions like feedback - it is sent like float value to last.fm right away.

@BusterNeece, is this something that can be fixed because I don't think that it's a LS issue.

**Reply by BusterNeece** [reply]
@gAlleb Hah, I know exactly what's causing this...mind creating a new issue for it? You can leave it sparse, but I'll get it fixed in parallel here.

**Reply by gAlleb** [reply]
Done! https://github.com/AzuraCast/AzuraCast/issues/7667

**Comment by VapidRealm** [comment]
This has stopped working for me. Had to remove the code to start my stations.

```
At /var/azuracast/stations/station/config/liquidsoap.liq, line 78 char 8 - line 84 char 1:
radio = audioscrobbler.submit(
  username="username",
...
radio
)

Error 14: Uncaught runtime error:
type: audioscrobbler,
message: "Error 4: {audioscrobbler=error(kind=\"audioscrobbler\",message=\"\").{trace=[], message=\"\", kind=\"audioscrobbler\"}, message=<fun>, kind=<fun>, failure=error(kind=\"failure\",message=\"\").{trace=[], message=\"\", kind=\"failure\"}, string=error(kind=\"string\",message=\"\").{trace=[], message=\"\", kind=\"string\"}, socket=error(kind=\"socket\",message=\"\").{trace=[], message=\"\", kind=\"socket\"}, output=error(kind=\"output\",message=\"\").{trace=[], message=\"\", kind=\"output\"}, not_found=error(kind=\"not_found\",message=\"\").{trace=[], message=\"\", kind=\"not_found\"}, json=error(kind=\"json\",message=\"\").{trace=[], message=\"\", kind=\"json\"}, invalid=error(kind=\"invalid\",message=\"\").{trace=[], message=\"\", kind=\"invalid\"}, http=error(kind=\"http\",message=\"\").{trace=[], message=\"\", kind=\"http\"}, file=error(kind=\"file\",message=\"\").{cross_device=error(kind=\"file.cross_device\",message=\"\").{trace=[], message=\"\", kind=\"file.cross_device\"}, trace=[], message=\"\", kind=\"file\"}, eval=error(kind=\"eval\",message=\"\").{trace=[], message=\"\", kind=\"eval\"}, clock=error(kind=\"clock\",message=\"\").{trace=[], message=\"\", kind=\"clock\"}, assertion=error(kind=\"assertion\",message=\"\").{trace=[], message=\"\", kind=\"assertion\"}, methods=<fun>, catch=<fun>, on_error=<fun>, raise=<fun>, register=<fun>}",
stack: at /var/azuracast/stations/station/config/liquidsoap.liq, line 78 char 8 - line 84 char 1,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 415 char 4 - line 420 char 5,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 85, char 2-37,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 59 char 2 - line 70 char 5,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 66 char 4 - line 69 char 5,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 66 char 4 - line 69 char 5,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 59 char 2 - line 70 char 5,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 59 char 2 - line 70 char 5,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 85, char 2-37,
at /usr/share/liquidsoap/libs/extra/audioscrobbler.liq, line 415 char 4 - line 420 char 5,
at /var/azuracast/stations/station/config/liquidsoap.liq, line 78 char 8 - line 84 char 1
```

**Comment by VapidRealm** [comment]
Just to follow up. With the code in, I get a SpawnException error, and Liquidsoap will not start. Rolling Release #85dce20
